tools/src/gencerts/Configuration.hs

```diff
@@ -0,0 +1,256 @@
+            let
+                clDN = mkDistinguishedName tlsClient
+                suffix = if i == 0 then "" else "_" <> show i
+            in CertDescription
+                { certConfiguration = tlsClient
+                , certSerial        = 3 + i
+                , certExtensions    = clExtensionsV3 clDN caDN
+                , certIssuer        = caDN
+                , certSubject       = clDN
+                , certGenKeys       = certGenKeys
+                , certSigningKey    = caPriv
+                , certOutDir        = outDirClients
+                , certFilename      = "client" <> suffix
+                , certChecks        = defaultChecks { checkFQHN = False }
+                }
+    in
+        (caConfig, svConfig : clConfigs)
+
+
+--
+-- INTERNALS / UTILS
+--
+
+caExtensionsV3 :: DistinguishedName -> [ExtensionRaw]
+caExtensionsV3 dn =
+    let
+        keyUsage         = ExtKeyUsage [KeyUsage_keyCertSign, KeyUsage_cRLSign]
+        basicConstraints = ExtBasicConstraints True (Just 0)
+        subjectKeyId     = ExtSubjectKeyId (hashDN dn)
+        authorityKeyId   = ExtAuthorityKeyId (hashDN dn)
+    in
+        [ extensionEncode True keyUsage
+        , extensionEncode True basicConstraints
+        , extensionEncode False subjectKeyId
+        , extensionEncode False authorityKeyId
+        ]
+
+
+usExtensionsV3 :: ExtKeyUsagePurpose -> DistinguishedName -> DistinguishedName -> [ExtensionRaw]
+usExtensionsV3 purpose subDN issDN =
+    let
+        keyUsage         = ExtKeyUsage [KeyUsage_digitalSignature, KeyUsage_keyEncipherment]
+        basicConstraints = ExtBasicConstraints False Nothing
+        subjectKeyId     = ExtSubjectKeyId (hashDN subDN)
+        authorityKeyId   = ExtAuthorityKeyId (hashDN issDN)
+        extendedKeyUsage = ExtExtendedKeyUsage [purpose]
+    in
+        [ extensionEncode True keyUsage
+        , extensionEncode False extendedKeyUsage
+        , extensionEncode False basicConstraints
+        , extensionEncode False subjectKeyId
+        , extensionEncode False authorityKeyId
+        ]
+
+
+svExtensionsV3 :: DistinguishedName -> DistinguishedName -> NonEmpty String -> [ExtensionRaw]
+svExtensionsV3 subDN issDN altNames =
+    let
+        subjectAltName = ExtSubjectAltName (AltNameDNS <$> NonEmpty.toList altNames)
+    in
+        extensionEncode False subjectAltName : usExtensionsV3 KeyUsagePurpose_ServerAuth subDN issDN
+
+
+clExtensionsV3 :: DistinguishedName -> DistinguishedName -> [ExtensionRaw]
+clExtensionsV3 =
+    usExtensionsV3 KeyUsagePurpose_ClientAuth
+
+
+aesonDropPrefix :: String -> Aeson.Options
+aesonDropPrefix pre = Aeson.defaultOptions
+    { Aeson.fieldLabelModifier = \s -> maybe s lowerFirst (stripPrefix pre s) }
+  where
+    lowerFirst :: String -> String
+    lowerFirst []    = []
+    lowerFirst (h:q) = Char.toLower h : q
+
+
+forEach :: [a] -> ((Int, a) -> b) -> [b]
+forEach xs fn =
+    zipWith (curry fn) [0..(length xs - 1)] xs
```

Comment:
This can be `zipWith (curry fn) [0..] xs`. `zipWith` will truncate the result list to the shorter of the two lists you give it, so the infinite list is :ok_hand: , and in fact, you can do `take 10 (zipWith (+) [0..] [1..])` and the laziness works out fine -- if you try to take `length [0..]`, then it will never terminate!

Reply:
:ok_hand: